HTTP Fields with OWS (Optional WhiteSpace) in value are not properly parsed in Jetty 12

**Jetty version(s)**
12.0.7

**Jetty Environment**
ee10

**Java version/vendor** `(use: java -version)`
C:\tools\jdk21\bin>java -version
openjdk version "21.0.2" 2024-01-16 LTS
OpenJDK Runtime Environment Corretto-21.0.2.13.1 (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.2.13.1 (build 21.0.2+13-LTS, mixed mode, sharing)

**OS type/version**
Windows 10 Enterprise Version 22H2

**Description**
Jetty 12's HttpParser throws an exception when parsing an HTTP response that contains multiple spaces before the Content-Length header value. This does not happen with Jetty 11, and the extra space is not in violation of the spec. The exception is: 

```
Caused by: o.e.j.h.BadMessageException: 400: Invalid Content-Length Value
	at o.e.j.h.HttpParser.convertContentLength(HttpParser.java:1138)
	at o.e.j.h.HttpParser.parsedHeader(HttpParser.java:991)
	at o.e.j.h.HttpParser.parseFields(HttpParser.java:1199)
	at o.e.j.h.HttpParser.parseNext(HttpParser.java:1545)
	... 21 common frames omitted
Caused by: j.l.NumberFormatException: null
	... 25 common frames omitted
```
**How to reproduce?**
The following JUnit tests reproduce/illustrate the problem:

```
package jetty;

import org.eclipse.jetty.http.HttpTester;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
* These two tests illustrate an issue in Jetty 12 with respect to parsing a
* response that has more than one space before the Content-Length header value
* (and after the colon).
*/
class JettyHttpParserTest {

       // This test passes in both Jetty 11 and Jetty 12
       // Note that there is only one space before the Content-Length header value in this test 
       @Test
       @DisplayName("Content-Length with no leading spaces")
       void ParseResponseWithNoLeadingSpaces() throws Exception {
              String response = """
              HTTP/1.1 200 OK
              Content-Length: 12
              Content-Type: text/plain
              Hello World!
              """;

              Assertions.assertDoesNotThrow(() -> HttpTester.parseResponse(response));
       }

       // This test passes in Jetty 11, but fails in Jetty 12
       // Note that there are two spaces in front of the Content-Length header value in this test
       @Test
       @DisplayName("Content-Length with leading spaces")
       void ParseResponseWithLeadingSpaces() throws Exception {
              String response = """
              HTTP/1.1 200 OK
              Content-Length:  12
              Content-Type: text/plain
              Hello World!
              """;

              Assertions.assertDoesNotThrow(() -> HttpTester.parseResponse(response));
       }
}

```

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
@@ -1228,11 +1228,21 @@ public class HttpParser
 
         for (int i = 0; i < length; i++)
         {
-            char c = valueString.charAt(i);
-            if (c < '0' || c > '9')
-                throw new BadMessageException("Invalid Content-Length Value", new NumberFormatException());
+            char ch = valueString.charAt(i);
+            HttpTokens.Token t = HttpTokens.getToken(ch);
 
-            value = Math.addExact(Math.multiplyExact(value, 10), c - '0');
+            switch (t.getType())
+            {
+                case SPACE:
+                case HTAB:
+                    // ignore OWS
+                    continue;
+                case DIGIT:
+                    value = Math.addExact(Math.multiplyExact(value, 10), ch - '0');
+                    break;
+                default:
+                    throw new BadMessageException("Invalid Content-Length Value", new NumberFormatException());
+            }
         }
         return value;
     }

--- a/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
+++ b/jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java
@@ -95,6 +95,7 @@ public class HttpParser
     public static final Logger LOG = LoggerFactory.getLogger(HttpParser.class);
     public static final int INITIAL_URI_LENGTH = 256;
     private static final int MAX_CHUNK_LENGTH = Integer.MAX_VALUE / 16 - 16;
+    private static final String UNMATCHED_VALUE = "\u0000";
 
     /**
      * Cache of common {@link HttpField}s including: <UL>
@@ -165,7 +166,7 @@ public class HttpParser
             Map<String, HttpField> map = new LinkedHashMap<>();
             for (HttpHeader h : HttpHeader.values())
             {
-                HttpField httpField = new HttpField(h, (String)null);
+                HttpField httpField = new HttpField(h, UNMATCHED_VALUE);
                 map.put(httpField.toString(), httpField);
             }
             return map;
@@ -1228,21 +1229,11 @@ public class HttpParser
 
         for (int i = 0; i < length; i++)
         {
-            char ch = valueString.charAt(i);
-            HttpTokens.Token t = HttpTokens.getToken(ch);
+            char c = valueString.charAt(i);
+            if (c < '0' || c > '9')
+                throw new BadMessageException("Invalid Content-Length Value", new NumberFormatException());
 
-            switch (t.getType())
-            {
-                case SPACE:
-                case HTAB:
-                    // ignore OWS
-                    continue;
-                case DIGIT:
-                    value = Math.addExact(Math.multiplyExact(value, 10), ch - '0');
-                    break;
-                default:
-                    throw new BadMessageException("Invalid Content-Length Value", new NumberFormatException());
-            }
+            value = Math.addExact(Math.multiplyExact(value, 10), c - '0');
         }
         return value;
     }
@@ -1387,25 +1378,28 @@ public class HttpParser
                                     String n = cachedField.getName();
                                     String v = cachedField.getValue();
 
-                                    if (CASE_SENSITIVE_FIELD_NAME.isAllowedBy(_complianceMode))
+                                    if (v != UNMATCHED_VALUE)
                                     {
-                                        // Have to get the fields exactly from the buffer to match case
-                                        String en = BufferUtil.toString(buffer, buffer.position() - 1, n.length(), StandardCharsets.US_ASCII);
-                                        if (!n.equals(en))
+                                        if (CASE_SENSITIVE_FIELD_NAME.isAllowedBy(_complianceMode))
                                         {
-                                            reportComplianceViolation(CASE_SENSITIVE_FIELD_NAME, en);
-                                            n = en;
-                                            cachedField = new HttpField(cachedField.getHeader(), n, v);
+                                            // Have to get the fields exactly from the buffer to match case
+                                            String en = BufferUtil.toString(buffer, buffer.position() - 1, n.length(), StandardCharsets.US_ASCII);
+                                            if (!n.equals(en))
+                                            {
+                                                reportComplianceViolation(CASE_SENSITIVE_FIELD_NAME, en);
+                                                n = en;
+                                                cachedField = new HttpField(cachedField.getHeader(), n, v);
+                                            }
                                         }
-                                    }
 
-                                    if (v != null && isHeaderCacheCaseSensitive())
-                                    {
-                                        String ev = BufferUtil.toString(buffer, buffer.position() + n.length() + 1, v.length(), StandardCharsets.ISO_8859_1);
-                                        if (!v.equals(ev))
+                                        if (isHeaderCacheCaseSensitive())
                                         {
-                                            v = ev;
-                                            cachedField = new HttpField(cachedField.getHeader(), n, v);
+                                            String ev = BufferUtil.toString(buffer, buffer.position() + n.length() + 1, v.length(), StandardCharsets.ISO_8859_1);
+                                            if (!v.equals(ev))
+                                            {
+                                                v = ev;
+                                                cachedField = new HttpField(cachedField.getHeader(), n, v);
+                                            }
                                         }
                                     }
 
@@ -1414,7 +1408,7 @@ public class HttpParser
 
                                     int posAfterName = buffer.position() + n.length() + 1;
 
-                                    if (v == null || (posAfterName + v.length()) >= buffer.limit())
+                                    if (v == UNMATCHED_VALUE || (posAfterName + v.length()) >= buffer.limit())
                                     {
                                         // Header only
                                         setState(FieldState.VALUE);

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
@@ -2023,12 +2023,84 @@ public class HttpParserTest
         assertEquals("GET", _methodOrVersion);
         assertEquals("/test", _uriOrStatus);
         assertEquals("HTTP/1.1", _versionOrReason);
-        assertEquals("localhost", _val[0]);
         assertEquals("Host", _hdr[0]);
         assertEquals("localhost", _val[0]);
 
         assertEquals(_content.length(), 10);
         assertEquals(parser.getContentLength(), 10);
+        assertTrue(_headerCompleted);
+        assertTrue(_messageCompleted);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        " chunked ",
+        "chunked ",
+        " chunked",
+        "\tchunked",
+        "\tchunked\t",
+        "chunked\t",
+        " \t \t \t chunked"
+    })
+    public void testTransferEncodingWithOWS(String transferEncoding)
+    {
+        String rawRequest = """
+            GET /test HTTP/1.1\r
+            Host: localhost\r
+            Transfer-Encoding: @TE@\r
+            \r
+            1\r
+            X\r
+            0\r
+            \r
+            """.replace("@TE@", transferEncoding);
+        ByteBuffer buffer = BufferUtil.toBuffer(rawRequest);
+
+        HttpParser.RequestHandler handler = new Handler();
+        HttpParser parser = new HttpParser(handler);
+        parseAll(parser, buffer);
+
+        assertEquals("GET", _methodOrVersion);
+        assertEquals("/test", _uriOrStatus);
+        assertEquals("HTTP/1.1", _versionOrReason);
+        assertEquals("Host", _hdr[0]);
+        assertEquals("localhost", _val[0]);
+        assertEquals("Transfer-Encoding", _hdr[1]);
+        assertEquals("chunked", _val[1]);
+
+        assertTrue(_headerCompleted);
+        assertTrue(_messageCompleted);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        " testhost ",
+        "testhost ",
+        " testhost",
+        "\ttesthost",
+        "\ttesthost\t",
+        "testhost\t",
+        " \t \t \t testhost"
+    })
+    public void testHostWithOWS(String host)
+    {
+        String rawRequest = """
+            GET /test HTTP/1.1\r
+            Host: @HOST@\r
+            \r
+            """.replace("@HOST@", host);
+        ByteBuffer buffer = BufferUtil.toBuffer(rawRequest);
+
+        HttpParser.RequestHandler handler = new Handler();
+        HttpParser parser = new HttpParser(handler);
+        parseAll(parser, buffer);
+
+        assertEquals("GET", _methodOrVersion);
+        assertEquals("/test", _uriOrStatus);
+        assertEquals("HTTP/1.1", _versionOrReason);
+        assertEquals("Host", _hdr[0]);
+        assertEquals("testhost", _val[0]);
+
         assertTrue(_headerCompleted);
         assertTrue(_messageCompleted);
     }

--- a/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
+++ b/jetty-core/jetty-http/src/test/java/org/eclipse/jetty/http/HttpParserTest.java
@@ -1971,7 +1971,8 @@ public class HttpParserTest
         "+10",
         "1.0",
         "1,0",
-        "10,"
+        "10,",
+        "10A"
     })
     public void testBadContentLengths(String contentLength)
     {
@@ -1992,6 +1993,44 @@ public class HttpParserTest
         parser.atEOF();
         parser.parseNext(BufferUtil.EMPTY_BUFFER);
         assertEquals(HttpParser.State.CLOSED, parser.getState());
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {
+        " 10 ",
+        "10 ",
+        " 10",
+        "\t10",
+        "\t10\t",
+        "10\t",
+        " \t \t \t 10"
+    })
+    public void testContentLengthWithOWS(String contentLength)
+    {
+        String rawRequest = """
+            GET /test HTTP/1.1\r
+            Host: localhost\r
+            Content-Length: @LEN@\r
+            \r
+            1234567890
+            """.replace("@LEN@", contentLength);
+        ByteBuffer buffer = BufferUtil.toBuffer(rawRequest);
+
+        HttpParser.RequestHandler handler = new Handler();
+        HttpParser parser = new HttpParser(handler);
+        parseAll(parser, buffer);
+
+        assertEquals("GET", _methodOrVersion);
+        assertEquals("/test", _uriOrStatus);
+        assertEquals("HTTP/1.1", _versionOrReason);
+        assertEquals("localhost", _val[0]);
+        assertEquals("Host", _hdr[0]);
+        assertEquals("localhost", _val[0]);
+
+        assertEquals(_content.length(), 10);
+        assertEquals(parser.getContentLength(), 10);
+        assertTrue(_headerCompleted);
+        assertTrue(_messageCompleted);
     }
 
     @ParameterizedTest

